### ISSUE TITLE
defrecord+ forces caller to load potemkin namespace

### ISSUE THREAD

**Issue by mrrodriguez** [Prismatic/schema#258]
I understand the motivation behind the schema library decoupling from the potemkin dependency so that consumers could opt-in to using it only if desired.

However, I find the usage of `schema.core/defrecord+` to be hard to work with now since it introduces a dependency on a namespace that is not "visible" when actually writing the code.

e.g.  If I make a namespace like:
```clj
(ns my.test.ns (:require [schema.core :as s]))

(s/defrecord+ MyTest [x :- s/Str])
```

Even if I have an explicit dependency in my project on potemkin, this namespace above will fail when JIT compiled unless some other namespace somewhere calls some sort of loading function on the potemkin namespace - typically this is via `:require`.

Technically, `my.test.ns` should be responsible for adding the `:require` on potemkin, but that isn't at all obvious when reading through the code.  In my situation, we have a lot of namespace that use `schema.core/defrecord` prior to the split to `schema.core/defrecord+`.  We still want the behavior of `schema.core/defrecord+` because we like the "reload-safe" semantics that `potemkin/defrecord+` provides.

However, now we would need to add a `:require` on potemkin all over the place in our files that use `schema.core/defrecord+` if we want to make each namespace be something that can "stand-alone" and ensure all its direct dependencies are loaded - JIT compiled here.

I've seen this issue come up writing macros before.  The macro wants to use a symbol from another namespace, but for some reason that namespace can't be explicitly loaded by a `:require` in the namespace that defines the macro.  This is a form of "lazy-loading" I'd call it.
However, I'd expect the macro to be written instead like:
```clj
(defmacro defrecord+
  "Like defrecord, but emits a record using potemkin/defrecord+.  You must provide
   your own dependency on potemkin to use this."
  {:arglists '([name field-schema extra-key-schema? extra-validator-fn? & opts+specs])}
  [name field-schema & more-args]
  (require 'potemkin) ; NEW LINE HERE
  (apply macros/emit-defrecord 'potemkin/defrecord+ &env name field-schema more-args))

```

Would this be an acceptable approach to take?  Is there any particular reason that we wouldn't want to do this instead?

**Comment by w01fe** [comment]
That seems fine to me -- PR welcome.  

A (possibly cleaner) alternative would be to make a new namespace `schema.potemkin` that requires `potemkin`, and move `s/defrecord+` there.  Thoughts?

**Comment by mrrodriguez** [comment]
I like the idea of another namespace for it just because it makes it clear there is a direct dependency on potemkin and the namespace (JIT) loading will just happen naturally when you `:require` it.
I can submit a PR soon for that.

**Comment by w01fe** [comment]
Great, thanks!  Maybe leave the existing one for now and mark it as deprecated with a note that it's moved to another namespace?  

Since the file is clj-only, it should go into the clj directory.  Files in the cljx directory produce both a clj and cljs file, and there's no point in producing an empty cljs namespace for this.  In the future it could always move to cljx to add cljs support without breaking users of the library.  

This will all go away in a while, once we feel it's safe to drop 1.6 support and can move to reader conditionals.

**Comment by mrrodriguez** [comment]
I've created the PR https://github.com/Prismatic/schema/pull/261.

Side note:  
This was happening against the master branch of schema, not included my PR changes.  So it is unrelated.
When I run the command:
```
lein all test
```

I get a `java.lang.OutOfMemoryError: Java heap space` from the 
`schema.experimental.generators-test/spec-test` because the project.clj for schema does not appear to specify any `:jvm-opts` to boost the memory.  I'm assuming that this isn't an issue to others because of having a default set in their .lein/profiles.clj.

However, I think it may make sense for the schema project.clj itself to declare what the `:jvm-opts` are expected to be to run the tests.  I didn't do that with this PR though because I wasn't sure what your thoughts on it are.

**Comment by w01fe** [comment]
Haven't seen the OOM issue before -- maybe it's a platform thing, I don't have any defaults set in my .lein/profiles.clj.  What platform / jvm version are you on?

**Comment by mrrodriguez** [comment]
I looked at this some more.  I have several JDK versions installed on my machine.  Due to this it was a little confusing.

With the 1.6 JVM I had this OOM issue.  With 1.7 (or higher) JVM I did not.  The reason for this is the default MaxHeapSize changed from 1.6 to 1.7:

```
java -XX:+PrintFlagsFinal -version | grep MaxHeapSize

Picked up JAVA_TOOL_OPTIONS: -Djava.awt.headless=true
    uintx MaxHeapSize                               = 132120576       {product}
java version "1.6.0_65"
Java(TM) SE Runtime Environment (build 1.6.0_65-b14-462-11M4609)
Java HotSpot(TM) 64-Bit Server VM (build 20.65-b04-462, mixed mode)
```

vs

```
java -XX:+PrintFlagsFinal -version| grep MaxHeapSize

Picked up JAVA_TOOL_OPTIONS: -Djava.awt.headless=true
    uintx MaxHeapSize                              := 2147483648      {product}
java version "1.7.0_71"
Java(TM) SE Runtime Environment (build 1.7.0_71-b14)
Java HotSpot(TM) 64-Bit Server VM (build 24.71-b01, mixed mode)
```

Java 6 is pretty old and had a low default heap space relative to Java 7, so I suppose this isn't really an issue to worry about much with the schema project.  The one advantage I have seen with setting the `:jvm-opts` in the Lein project.clj of a project is to be able to keep a closer watch over any sudden jumps in memory usage that come when things are changed/added.

**Comment by w01fe** [comment]
Thanks for the follow-up -- Java 6 (and even Java 7 now I believe) is EOL so I'm not too worried about it.    Tracking memory is interesting but I think we're probably doing that more effectively in other places (e.g., testing on our whole codebase) so open to it but not sure there's a lot of value in adding it to the jvm opts here.

**Comment by mrrodriguez** [comment]
This is fine to me.  It was just a little sneaky.  If you have alternative methods of tracking on memory usage over time, that sounds good.

I could work around it by running `lein` with a bit more options from the cmd line anyways:
```
lein update-in : assoc :jvm-opts '["-Xmx2G"]' -- all test
```

**Comment by w01fe** [comment]
Fixed by #261 (thanks!)
